### ISSUE TITLE
How to set the timestep？I want to compare the effect of ESN and LSTM in a time series prediction task

### ISSUE THREAD

**Issue by imhithanks** [TUD-STKS/PyRCN#6]
I want to preditct  the daily price of  a stock with several features and compare the prediction effect of ESN and LSTM .
This example  use the price of  today to predict the price of  tomorrow.(timestep=1）
https://github.com/TUD-STKS/PyRCN/blob/master/examples/stock-price-prediction.ipynb
X : ndarray of shape (n_samples, n_features)

Using LSTM, we can use the price in last 10 days to predict the price of tomorrow.(n_timestep=10）
 In  tensorflow2.0 ,we reshape the input X into (n_samples, n_timesteps，n_features)

1）How can I set  the n_timestep in PyRCN like LSTM？
（I want to  use the price and it's features  in last 10 days to predict the price of tomorrow.）
2）How can I set the parameters to compare ESN and LSTM？
Should I use the same  number of neurons？

Thank you for your help.

**Comment by renierts** [comment]
Hi imhithanks,

1) As you have noticed, X is of shape (n_samples, n_features). If you would like to use the price of the last 10 days (n_timesteps=10), you could create X of shape (n_samples, n_features*n_timesteps), where each sample contains then the price of the last 10 days. I think, numpy provides stride operations, which can be used to create X easily.

Note that you should increase K_in for a multidimensional input. For you, K_in = 5 might be a good choice. Afterwards, you should optimize the hyper-parameters as described in the digits notebook.

2) You have several opportunities. What I usually do is comparing the trainable parameters of ESNs ((number of neurons + 1)* number of outputs) and other models (LSTM, MLP etc.).

**Comment by imhithanks** [comment]
There's still something  I don't quite understand.
For example,  There are daily price of a stock and it's  9 features in 1000 days . n_timestep=10 .
I  create X of shape (1000, 9*10)，Y of shape (1000, 1)
1)K_in should be 9？
param k_in: This element represents the sparsity of the connections between the input and recurrent nodes. It determines the number of features that every node inside the reservoir receives.
I don't understand this parameter. Is it right that k_in=5 means the other 4 features are useless?
2) I want to compare the prediction effect of ESN and LSTM.
I  adjust there parameters on the validation set respectively and compare their MSE on test set. 
Is it a right way?

**Comment by renierts** [comment]
1) K_in = 5 would mean the each neuron inside the reservoir receives 5 randomly selected features. Every neuron thus receives different combinations of 5 features. 

2) Yes, this is a right way to go.

**Comment by imhithanks** [comment]
@renierts
 I got it. There are still several problems.
1)Can I set bi_directional=True in  a time series prediction task？Is there any reference about it？
If I make bi_directional=True, Can I call it Bi-ESN？
 bi_directional: If True, the input sequences are passed through the network two times, forward and backward.

2) Must the number of parameters be less than the number of samples
For example , I just have 1000 samples .Can I set reservoir_size=1500？

3）Does different reservoir_activation can make a big difference in prediction effect ?
reservoir_activation: {‘tanh’, ‘identity’, ‘logistic’, ‘relu’}

4) I hope ESN can perform better than LSTM. But ESN is not better than LSTM on my stock dataset.
From your experience, Does ESN often perform worse than LSTM？
I'm very anxious and don't know how to improve the prediction effect of ESN.

Thank you for your help and patience ！！

**Comment by renierts** [comment]
Hi imhithanks,

1) This is strongly task-dependent. As I am not frequently working with time-series prediction, I cannot give you a clear reference. Regarding "Bi-ESN", you can call it that way.
2) I would strongly recommend to try out various reservoir sizes, e.g. from 50 up to 1000 samples. This is strongly task-dependent.
3) This is also task-dependent and you should compare the impact of activation functions (as in any neural network).
4) From my experience with music, ESNs performed similar or even a bit better than LSTMs. You should try to find some reference publications that utilized ESNs for stock price prediction.

**Comment by imhithanks** [comment]
@renierts  
There are still something i want to learn from you.
1)  I wonder how to explain "the input sequences are passed through the network two times "mathematically.
The Wout is decided when the  sequences are passed through . How does it change when the sequences are passed through backward？
( If i set bi_directional=True,the prediction MSE reduces largely on test set. I wonder if it use the future information )

2) input_scaling: This element represents the input scaling factor from the input to the reservoir. 
 For example , input_scaling=5 means that the randomly generated matrix Win is multiplied by 5. Is it right?

3) In the example below , n_feature=1. I found that I can set K_in=4.What is the other 3 feature?
https://github.com/TUD-STKS/PyRCN/blob/master/examples/stock-price-prediction.ipynb
X : ndarray of shape (n_samples,1)

**Comment by renierts** [comment]
1) Yes, the sequence is passed through the ESN forward and backward if you set bi_directional=True. Thus, it utilizes future information.
2) Yes
3) K_in does not affect the number of features. The result should not strongly change if you increase K_in that example. In the next version of PyRCN, which I have already started, I will throw an error if K_in > n_features.

**Comment by imhithanks** [comment]
@renierts
I see. Thank you for your help！！

**Comment by renierts** [comment]
You're welcome
